sources/engine/Stride.Rendering/Rendering/Mesh.cs

```diff
@@ -76,5 +82,165 @@ public Mesh(MeshDraw meshDraw, ParameterCollection parameters)
+
+       public Matrix[] MATBSHAPE { get; set; }
+
+       public float BasisKeyWeight { get; set; }
+
+       public void ProcessBlendShapes()
+       {
+            if (BlendShapeProcessingNecessary)
+            {
+                if (Shapes == null || Shapes.Count < 1) { return; }
+                var blendShapeVertices = AdjustPositionToDrawInstance();
+                var blendShapesWeights = GetBlendWeights(out float cummulativeWeight); ;
+                BasisKeyWeight = 1 - cummulativeWeight;
+
+                MATBSHAPE = new Matrix[Draw.VertexMapping.Length*blendShapesWeights.Length];
+                for (int iBendShape = 0; iBendShape < blendShapesWeights.Length; iBendShape++)
+                {
+                    for (int iVert = 0; iVert < Draw.VertexMapping.Length; iVert++)
+                    {
+                        var vectorValue = blendShapeVertices.ElementAt(iBendShape * Draw.VertexMapping.Length + iVert);
+                        MATBSHAPE[iBendShape * Draw.VertexMapping.Length + iVert].Row1 = new Vector4(vectorValue, blendShapesWeights[iBendShape]);
+
+                    }
+                }
+                BlendShapeProcessingNecessary = false;
+            }
+        }
+        
+
+        public void AddBlendShapes(Shape shape, float weight)
+        {
+            (Shapes ??= new()).Add(shape, weight); 
+            Parameters.Set(MaterialKeys.HasBlendShape, true);
+            
+        }
+
+        public float[] GetBlendWeights(out float cummulativeWeight)
+        {
+            float[] shapeWeights = new float[Shapes.Count];
+            cummulativeWeight = 0f;
+             
+            for (int i = 0; i < Shapes.Count; i++)
+            {
+                var shapeKV = Shapes.ElementAt(i);
+                var shape = shapeKV.Key;
+                var shapeWeight = Math.Clamp(shapeKV.Value, 0f, 1f);
+                float adjustedWeight = shapeWeight;
+               // if (cummulativeWeight >= 1f) { adjustedWeight = 0f; }
+               // else if (cummulativeWeight + shapeWeight > 1) { adjustedWeight = 1 - cummulativeWeight; }
+                //else
+                //{
+                  //  adjustedWeight = shapeWeight;
+                //}
+                cummulativeWeight += adjustedWeight;
+                shapeWeights[i] = adjustedWeight;
+            }
+            return shapeWeights;
+        }
+
+
+  
+        public List<Vector3> AdjustPositionToDrawInstance()
+        {
+            List<Vector3> adjustedPositons = new List<Vector3>();
+            foreach (var shape in Shapes)
+            {
+                var posBlend = shape.Key.Position;
+                List<int> originalVerticesIDS = Draw.VCPOLYIN.Select(c => c.Item2).ToList();
+                Dictionary<int, Vector3> mappings = new Dictionary<int, Vector3>();
+                foreach (var tup_id_vec in Draw.VCPOLYIN)
+                {
+                    if (!mappings.ContainsKey(tup_id_vec.Item2))
```

Comment:
use the new TryAdd of the dictionary to avoid double lookup in the dictionary

Reply:
Done!